src/adaptors/adaptor_tls.c

```diff
@@ -0,0 +1,645 @@
+               QD_LOG_ERROR,
+               "[C%" PRIu64 "] Error processing TLS: %s, tls_has_output=%i",
+               tls->conn_id,
+               error_msg,
+               tls->tls_has_output);
+        return false;
+    }
+    return true;
+}
+
+static void take_back_input_encrypt_buffs(qd_tls_t *tls)
+{
+    pn_raw_buffer_t take_incoming_buf;
+    size_t          take_input_count;
+
+    while ((take_input_count = pn_tls_take_encrypt_input_buffers(tls->tls_session, &take_incoming_buf, 1))) {
+        assert(take_input_count == 1);
+        (void) take_input_count;  // prevent unused variable warning
+        qd_adaptor_buffer_t *adaptor_buff = (qd_adaptor_buffer_t *) take_incoming_buf.context;
+        free_qd_adaptor_buffer_t(adaptor_buff);
+    }
+}
+
+/**
+ * Takes all decrypt input buffers back from tls by repeatedly calling pn_tls_take_decrypt_input_buffers
+ * in a while loop. Frees the underlying adaptor buffer.
+ */
+static void take_back_input_decrypt_buffs(qd_tls_t *tls)
+{
+    pn_raw_buffer_t take_incoming_buf;
+    size_t          take_input_count;
+    while ((take_input_count = pn_tls_take_decrypt_input_buffers(tls->tls_session, &take_incoming_buf, 1))) {
+        assert(take_input_count == 1);
+        (void) take_input_count;  // prevent unused variable warning
+        qd_adaptor_buffer_t *adaptor_buff = (qd_adaptor_buffer_t *) take_incoming_buf.context;
+        free_qd_adaptor_buffer_t(adaptor_buff);
+    }
+}
+
+size_t get_pn_raw_buffer_fetch_size(qd_tls_t *tls)
+{
+    size_t tls_input_buff_capacity = pn_tls_get_decrypt_input_buffer_capacity(tls->tls_session);
+    size_t fetch_size              = MIN(tls_input_buff_capacity, TLS_MAX_INPUT_CAPACITY);
+    return fetch_size;
+}
+
+int qd_tls_decrypt(qd_tls_t                 *tls,
+                   pn_raw_connection_t      *pn_raw_conn,
+                   qd_adaptor_buffer_list_t *decrypted_buffs,
+                   qd_adaptor_buffer_list_t *granted_read_buffs)
+{
+    if (tls->tls_error)
+        return QD_TLS_ERROR;
+
+    int encrypted_bytes_in = 0;
+
+    // How many decrypt input buffers can TLS take ?
+    size_t fetch_size = get_pn_raw_buffer_fetch_size(tls);
+
+    // Do not proceed if TLS has no decrypt input capacity.
+    if (fetch_size == 0) {
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%" PRIu64 "] qd_tls_decrypt fetch_size=0, returning", tls->conn_id);
+        return 0;
+    }
+
+    pn_raw_buffer_t raw_buffers[TLS_MAX_INPUT_CAPACITY];
+    size_t          n;
+    while ((n = pn_raw_connection_take_read_buffers(
+                pn_raw_conn,
+                raw_buffers,
+                fetch_size))) {  // Repeatedly get as much buffers (from raw connection) as possible
+        for (size_t i = 0; i < n; ++i) {
+            // Give each decrypt input buffer to proton tls
+            qd_adaptor_buffer_t *adaptor_buff = (qd_adaptor_buffer_t *) raw_buffers[i].context;
+            DEQ_REMOVE(*granted_read_buffs, adaptor_buff);
```

Comment:
So the HTTP/1 adaptor now has to start tracking read buffers granted to the tls layer?  Why?

Reply:
I am researching this and will let you know what I find